### PR TITLE
Incremental PushContext initilization

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
Based on @rshriram's pr #16076, to implement incremental push context init.  Basically it is to make use of the old push context as much as possible. For example,  if a service updated, only reinitialize its related configs, services and sidecarscopes. 

It is to improve the performance and scalability of the control plane. 

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[X] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

### CODE REVIEW

File: pilot/pkg/proxy/envoy/v2/discovery.go [hunk 1]
```diff
@@ -155,7 +155,13 @@ func NewDiscoveryServer(
 	}
 
 	// Flush cached discovery responses whenever services configuration change.
-	serviceHandler := func(*model.Service, model.Event) { out.clearCache(&model.PushRequest{Full: true}) }
+	serviceHandler := func(*model.Service, model.Event) {
+		pushReq := &model.PushRequest{
+			Full:               true,
+			ConfigTypesUpdated: map[string]struct{}{schemas.ServiceEntry.Type: {}},
```
Comment (howardjohn) [inline pilot/pkg/proxy/envoy/v2/discovery.go]: I don't think this is ServiceEntry only is it?

File: pilot/pkg/model/push_context.go [hunk 2]
```diff
@@ -686,55 +688,158 @@ func (ps *PushContext) SubsetToLabels(proxy *Proxy, subsetName string, hostname
...
+
+	// Must be initialized in the end
+	// Sidecars need to be updated if services, virtual services, destination rules, or the sidecar configs change
+	if servicesChanged || virtualServicesChanged || destinationRulesChanged || sidecarsChanged {
+		if err := ps.initSidecarScopes(env); err != nil {
+			return err
+		}
+	}
 
-	ps.initDone = true
```
Comment (howardjohn) [inline pilot/pkg/model/push_context.go]: where do we use this/set this now?

File: pilot/pkg/proxy/envoy/v2/discovery.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pilot/pkg/proxy/envoy/v2/discovery.go]: we use SE to represent service here

File: pilot/pkg/model/push_context.go (hunk 2 above)
Comment (hzxuzhonghu) [reply pilot/pkg/model/push_context.go]: will have a look

**Comment by hzxuzhonghu** [comment]
>did you do any perf testing if this?

no, i do not have the env as you did before with a great visualization.

File: pilot/pkg/model/push_context.go (hunk 2 above)
Comment (hzxuzhonghu) [reply pilot/pkg/model/push_context.go]: Still in the end of `InitContext`

File: pilot/pkg/model/push_context.go [hunk 3]
```diff
@@ -230,7 +232,7 @@ func (first *PushRequest) Merge(other *PushRequest) *PushRequest {
 	}
 
 	// Merge the target namespaces
-	if len(first.NamespacesUpdated) > 0 || len(other.NamespacesUpdated) > 0 {
+	if len(first.NamespacesUpdated) > 0 && len(other.NamespacesUpdated) > 0 {
```
Comment (rshriram) [inline pilot/pkg/model/push_context.go]: why did this change?

Comment (hzxuzhonghu) [reply pilot/pkg/model/push_context.go]: I think it was a bug. When one is empty, we should treat it as all namespaces updated.

Maybe i should do this in a separate pr.

File: pilot/pkg/proxy/envoy/v2/discovery.go (hunk 1 above)
Comment (rshriram) [reply pilot/pkg/proxy/envoy/v2/discovery.go]: what about Consul?

**Pull request merged**